File: hints/fullHouse.fs
module hints.fullHouse

(* Full House means:
 For a house there is only one cell that is neither given nor set i.e. has candidates *)

open core.sudoku
open core.puzzlemap
open core.hints

let fullHousePerHouse (p : puzzleMap) (cellCandidates : cellCandidates) (primaryHouse : house) : hintDescription option =

    let hhs =
        p.houseCellCandidateReductions primaryHouse cellCandidates
        |> CandidateReductions.filter (fun cr -> Digits.count cr.candidates > 0) 

    if CandidateReductions.count hhs = 1 then 
        let h = hhs.data.data.Head
        let cell = h.cell
        let candidate = first h.candidates

        let setCellValue = makeValue cell candidate

        Some { hintDescription.primaryHouses = Houses.singleton primaryHouse
               secondaryHouses = Houses.empty
               candidateReductions = CandidateReductions.empty
               setCellValueAction = Some setCellValue
               pointers = CandidateReductions.empty
               focus = Digits.empty }
    else None

let fullHouses (p : puzzleMap) (cellCandidates : cellCandidates) : hintDescription list =
    p.houses
    |> List.choose (fullHousePerHouse p cellCandidates)

File: src/fullHouse.ml
(* Full House means:
 For a house there is only one cell that is neither given nor set i.e. has candidates *)

open Sudoku
open Puzzlemap
(*F# open FSharp.Compatibility.OCaml F#*)

let fullHousePerHouse (p : puzzleMap) (cellCandidates : cellCandidates) (primaryHouse : house) : Hint.description option =

    let hhs =
        p.houseCellCandidateReductions primaryHouse cellCandidates
        |> List.filter (fun cr -> Digits.count cr.candidates > 0) 
        in

    if List.length hhs = 1 then 
        let h = List.hd hhs in
        let cell = h.cell in
        let candidate = Digits.first h.candidates in

        let setCellValue = Value.make cell candidate in

        Some { primaryHouses = Houses.singleton primaryHouse;
               secondaryHouses = Houses.empty;
               candidateReductions = [];
               setCellValueAction = Some setCellValue;
               pointers = [];
               focus = Digits.empty }
    else None

let find (p : puzzleMap) (cellCandidates : cellCandidates) : Hint.description list =
    p.houses
    |> Sset.choose (fullHousePerHouse p cellCandidates)
